pycontrails/models/cocipgrid/cocip_grid.py

```diff
@@ -212,47 +200,105 @@ def eval(
                 self.source.attrs[f"humidity_scaling_{k}"] = v
 
         self._parse_verbose_outputs()
-        self.attach_timedict()
+
+        self._set_timesteps()
         pbar = self.init_pbar()
 
-        summary: VectorDataset | None
-        summary_by_met_slice: list[VectorDataset] = []
-        existing_vectors: Iterator[GeoVectorDataset] = iter(())
+        met: MetDataset | None = None
+        rad: MetDataset | None = None
+
+        ef_summary: list[VectorDataset] = []
         verbose_dicts: list[dict[str, pd.Series]] = []
         contrail_list: list[GeoVectorDataset] = []
-        for time, filt in self.timedict.items():
-            met, rad = self._load_met_slices(time, pbar)
+        existing_vectors: Iterator[GeoVectorDataset] = iter(())
 
-            new_vectors = self._generate_new_vectors(filt)
-            evolved = []
-            for vector in new_vectors:
-                evolved.append(_evolve_vector(vector, met, rad, self.params, True, pbar))
-            for vector in existing_vectors:
-                evolved.append(_evolve_vector(vector, met, rad, self.params, False, pbar))
-            if not evolved:
-                break
+        for time_idx, time_end in enumerate(self.timesteps):
+            met, rad = self._maybe_downselect_met_rad(met, rad, time_end)
+
+            evolved_this_step = []
+            ef_summary_this_step = []
+
+            for vector in self._generate_new_vectors(time_idx):
+                res = _evolve_vector(
+                    vector,
+                    met=met,
+                    rad=rad,
+                    params=self.params,
+                    t=time_end,
+                    run_downwash=True,
+                    pbar=pbar,
+                )
+                if self.params["verbose_outputs_evolution"] and res.downwash:
+                    contrail_list.append(res.downwash)
+                if res.ef_summary:
+                    evolved_this_step.append(res.contrail)
+                    ef_summary_this_step.append(res.ef_summary)
+                    if self.params["verbose_outputs_evolution"]:
+                        contrail_list.append(res.contrail)
+                if self.params["verbose_outputs_formation"] and res.verbose_dict:
+                    verbose_dicts.append(res.verbose_dict)
 
-            vectors, summary_data, verbose_dicts_this_step, contrail_lists = zip(*evolved)
-            existing_vectors = combine_vectors(vectors, self.params["target_split_size"])
-            summary = VectorDataset.sum([r for r in summary_data if r])
-            summary_by_met_slice.append(summary)
+            for vector in existing_vectors:
+                res = _evolve_vector(
```

Comment:
Another question that might be obvious to pycontrails devs, but why are there two evolution routines?

Reply:
Nope, it's definitely not obvious! Part if this is some lingering history that could be refactored to improve. Any new waypoints that haven't yet undergone any advection need to be initialized with the CoCiP downwash routine. Right now, this `_evolve_vector` function is responsible for both the downwash and the evolution. The two calls are for the new waypoints (for which the downwash has not yet been calculated), and for the partially-evolved waypoints. It might be a little cleaner to implement this different. I'll think about that.